### PR TITLE
Put the path choices in clients instead of spec-wide

### PR DESCRIPTION
**Pull request by cwgoes**
Closes https://github.com/cosmos/ics/issues/288

### CODE REVIEW

**Comment by cwgoes** [comment]
- [ ] Update example code in ICS 2
- [ ] Update connection call proxy

**Comment by cwgoes** [comment]
Note that this also makes the canonical encoding client-type-specific, which I think is desirable.

**Comment by cwgoes** [comment]
Merging since we concurred in the meeting & I want to avoid conflicts. Glad to walk anyone through.

File: spec/ics-003-connection-semantics/README.md [hunk 1]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
-```
 
-```typescript
-function verifyNonMembership(
+function verifyConnectionState(
+  connection: ConnectionEnd,
+  height: uint64,
+  proof: CommitmentProof,
+  connectionIdentifier: Identifier,
+  connectionEnd: ConnectionEnd) {
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]: why are here 2 connection ends?

File: spec/ics-003-connection-semantics/README.md [hunk 2]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
+    client = queryClient(connection.clientIdentifier)
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]: should be client state

File: spec/ics-003-connection-semantics/README.md [hunk 3]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
+    return client.verifyConnectionState(connection, height, connection.counterpartyPrefix, proof, connectionIdentifier, connectionEnd)
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]: should be 

```go
client.verifyConnectionState(clientState, height, connection.counterpartyPrefix, proof, connection.Identifier, connection)
```

File: spec/ics-003-connection-semantics/README.md [hunk 4]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
 }
-```
 
-```typescript
-function verifyNonMembership(
+function verifyConnectionState(
+  connection: ConnectionEnd,
+  height: uint64,
+  proof: CommitmentProof,
+  connectionIdentifier: Identifier,
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]: is this required?

File: spec/ics-003-connection-semantics/README.md [hunk 5]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
-function verifyMembership(
+function verifyClientConsensusState(
   connection: ConnectionEnd,
   height: uint64,
   proof: CommitmentProof,
-  path: Path,
-  value: Value): bool {
+  clientIdentifier: Identifier,
+  consensusState: ConsensusState) {
     client = queryClient(connection.clientIdentifier)
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]: ditto, should be client state

File: spec/ics-003-connection-semantics/README.md [hunk 6]
```diff
@@ -134,29 +134,86 @@ function removeConnectionFromClient(
...
-    client.verifyMembership(height, proof, applyPrefix(connection.counterpartyPrefix, path), value)
+    return client.verifyClientConsensusState(connection, height, connection.counterpartyPrefix, proof, clientIdentifier, consensusState)
```
Comment (fedekunze) [inline spec/ics-003-connection-semantics/README.md]:
```go
return client.verifyClientConsensusState(clientState, height, connection.counterpartyPrefix, proof, clientState.Identifier, consensusState)
```

File: spec/ics-002-client-semantics/README.md [hunk 7]
```diff
@@ -473,27 +597,97 @@ function checkValidityAndUpdateState(
...
-function verifyNonMembership(
+function verifyChannelState(
   clientState: ClientState,
+  height: uint64,
+  prefix: CommitmentPrefix,
+  proof: CommitmentProof,
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  channelEnd: ChannelEnd) {
+    path = applyPrefix(prefix, "ports/{portIdentifier}/channels/{channelIdentifier}")
```
Comment (fedekunze) [inline spec/ics-002-client-semantics/README.md]: question: why do we want to expose the paths hardcoded on the client?

File: spec/ics-003-connection-semantics/README.md (hunk 1 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: There aren't. I'm not sure what you are referring to. Where is the other end?

Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: The second `connectionEnd` is the part actually being verified.

The first is the local connection.

File: spec/ics-003-connection-semantics/README.md (hunk 2 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: Why? The idea is that the client exposes this function.

File: spec/ics-003-connection-semantics/README.md (hunk 3 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: See above comment.

File: spec/ics-003-connection-semantics/README.md (hunk 4 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: Yes, we need to know the identifier to verify under.

File: spec/ics-003-connection-semantics/README.md (hunk 5 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: See above comment.

File: spec/ics-003-connection-semantics/README.md (hunk 6 above)
Comment (cwgoes) [reply spec/ics-003-connection-semantics/README.md]: See above comment.

File: spec/ics-002-client-semantics/README.md (hunk 7 above)
Comment (cwgoes) [reply spec/ics-002-client-semantics/README.md]: These are example implementations, not necessarily required.

They will vary per client type. See https://github.com/cosmos/ics/issues/286

**Pull request merged**